File: resume-app/src/main/java/a/gleb/resume_app/controller/ResumeController.java
package a.gleb.resume_app.controller;

import a.gleb.resume_app.controller.docs.ResumeSwagger;
import a.gleb.resume_app.model.request.ResumeRequest;
import a.gleb.resume_app.model.response.ResumeResponse;
import a.gleb.fellow_worker.http.response.PageResponse;
import a.gleb.resume_app.service.ResumeService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/resumes")
public class ResumeController implements ResumeSwagger {

    private final ResumeService resumeService;

    @Override
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ResumeResponse create(@Valid @RequestBody ResumeRequest request) {
        return resumeService.create(request);
    }

    @Override
    @GetMapping("/{id}")
    public ResumeResponse findById(@PathVariable UUID id) {
        return resumeService.findById(id);
    }

    @Override
    @GetMapping
    public PageResponse<ResumeResponse> findAll(@PageableDefault(size = 20) Pageable pageable) {
        var p = resumeService.findAll(pageable);
        return new PageResponse<>(p.getContent(), p.getNumber(), p.getSize(), p.getTotalElements(), p.getTotalPages());
    }

    @Override
    @GetMapping("/my")
    public List<ResumeResponse> findMy() {
        return resumeService.findMy();
    }

    @Override
    @PutMapping("/{id}")
    public ResumeResponse update(@PathVariable UUID id, @Valid @RequestBody ResumeRequest request) {
        return resumeService.update(id, request);
    }

    @Override
    @PostMapping(value = "/{id}/photo", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResumeResponse uploadPhoto(@PathVariable UUID id, @RequestParam("file") MultipartFile file) {
        return resumeService.uploadPhoto(id, file);
    }

    @Override
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable UUID id) {
        resumeService.delete(id);
    }
}

File: fellow-worker-service/src/main/kotlin/a/gleb/fellowworkerservice/controller/ResumeController.kt
/*
 * Copyright (c) 12-3/7/23, 10:06 PM
 * Created by https://github.com/alwayswanna
 */

package a.gleb.fellowworkerservice.controller

import a.gleb.apicommon.fellowworker.model.request.resume.ResumeApiModel
import a.gleb.apicommon.fellowworker.model.request.resume.SearchResumeApiModel
import a.gleb.apicommon.fellowworker.model.response.FellowWorkerResponseModel
import a.gleb.fellowworkerservice.configuration.OAUTH2_SECURITY_SCHEMA
import a.gleb.fellowworkerservice.service.ResumeService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.web.bind.annotation.*
import java.util.*

const val EMPLOYEE_TAG_NAME = "employee.tag.name"

@RestController
@RequestMapping("/api/v1/employee")
@Tag(name = EMPLOYEE_TAG_NAME)
class ResumeController(
    private val resumeService: ResumeService
) {

    @Operation(
        summary = "Создать резюме",
        security = [SecurityRequirement(name = OAUTH2_SECURITY_SCHEMA)]
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @PostMapping("/create-resume")
    suspend fun createResume(@RequestBody @Valid request: ResumeApiModel): FellowWorkerResponseModel {
        return resumeService.createResume(request)
    }

    @Operation(
        summary = "Редактировать резюме",
        security = [SecurityRequirement(name = OAUTH2_SECURITY_SCHEMA)]
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @PutMapping("/edit-resume")
    suspend fun editResume(@RequestBody @Valid request: ResumeApiModel): FellowWorkerResponseModel {
        return resumeService.editResume(request)
    }

    @Operation(
        summary = "Удалить резюме",
        security = [SecurityRequirement(name = OAUTH2_SECURITY_SCHEMA)]
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @DeleteMapping("/delete-resume")
    suspend fun deleteResume(@NotNull @RequestParam id: UUID): FellowWorkerResponseModel {
        return resumeService.deleteResume(id)
    }

    @Operation(
        summary = "Получить резюме по авторизованного пользователя",
        security = [SecurityRequirement(name = OAUTH2_SECURITY_SCHEMA)]
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @GetMapping("/current-user-resume")
    suspend fun getCurrentUserResume(): FellowWorkerResponseModel {
        return resumeService.getCurrentUserResume()
    }

    @Operation(
        summary = "Получить все резюме"
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @GetMapping("/get-all-resume")
    suspend fun getAllResume(): FellowWorkerResponseModel {
        return resumeService.allResume()
    }

    @Operation(
        summary = "Получить резюме по ID"
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @GetMapping("/get-resume-id")
    suspend fun getSingleResume(@NotNull @RequestParam id: UUID): FellowWorkerResponseModel {
        return resumeService.findResumeModelById(id)
    }

    @Operation(
        summary = "Поиск резюме по определенным фильтрам"
    )
    @ApiResponses(
        value = [
            ApiResponse(
                description = "OK", responseCode = "200",
                content = [Content(
                    mediaType = APPLICATION_JSON_VALUE,
                    schema = Schema(implementation = FellowWorkerResponseModel::class)
                )]
            ),
            ApiResponse(description = "Bad request", responseCode = "400"),
            ApiResponse(description = "Internal server error", responseCode = "500")
        ]
    )
    @PostMapping("/filter-resumes")
    suspend fun filterResumes(@RequestBody @Valid request: SearchResumeApiModel): FellowWorkerResponseModel{
        return resumeService.filter(request)
    }
}